app.py

```diff
@@ -0,0 +1,123 @@
+
+@app.route("/")
+@login_required
+def index():
+    return render_template("index.html", nickname=session['nickname'])
+
+@app.route("/login", methods=["GET", "POST"])
+def login():
+    """Log user in"""
+
+    # Forget any user_id
+    session.clear()
+
+    # User reached rout via POST (tried to log in via the form)
+    if request.method == "POST":
+
+        # Ensure username was submitted
+        if not request.form.get("username"):
+            # Info on 403: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/403
+            return apology("must provide username", 403)
+
+        elif not request.form.get("password"):
+            return apology("must provide password", 403)
+
+        # Query database for username
+        rows = db.execute("SELECT * FROM users WHERE username = ?", request.form.get("username"))
+
+        # Ensure username exists and password is correct
+        if len(rows) != 1 or not check_password_hash(rows[0]["hash"], request.form.get("password")):
+            return apology("invalid username and/or password", 403)
+
+        # Remeber what user logged in
+        session["user_id"] = rows[0]["user_id"]
+
+        # Remember user nickname
+        session["nickname"] = rows[0]["nickname"]
+
+        # Redirect user to home page
+        return redirect("/")
+
+    # User reached route via GET (clicked a link or redirected)
+    else:
+        return render_template("login.html")
+
+@app.route("/logout")
+def logout():
+    """Log user out"""
+
+    # Forget any user_id
+    session.clear()
+
+    # Redirect user to login form
+    return redirect("/")
+
+@app.route("/register", methods=["GET", "POST"])
+def register():
+    if request.method == "POST":
+        # ENSURE USERNAME WAS SUBMITTED
+        if not request.form.get("username"):
+            return apology("must provide username", 400)
+
+        # Ensure username does not exist already
+        username = request.form.get("username")
+        if db.execute("SELECT username FROM users WHERE username=(?)", username):
+            return apology("username already exists", 400)
+
+
+        # Ensure password is entered
+        if not request.form.get("password"):
+            return apology("must provide a password", 400)
+
+
+        # Ensure confirmation exists
+        if not request.form.get("confirmation"):
+            return apology("must provide a confirmation", 400)
+
+
+        # Ensure password and cinfirmation match
+        if request.form.get("password") != request.form.get("confirmation"):
+            return apology("Passwords do not match!", 400)
+
+
+        # Register user if they passed previous tests
+        password = request.form.get("password")
+        nickname = request.form.get("nickname")
+
+        db.execute("INSERT INTO users(username, hash, nickname) VALUES(?, ?, ?)", username, generate_password_hash(password), nickname)
+        return redirect("/")
+
+    else:
+        return render_template("register.html")
+
```

Comment:
I think flask_session could be added to .gitignore, I'll do that, since we don't have a .gitignore yet
```
/flask_session
```